kuma/scrape/sources/document.py

```diff
@@ -0,0 +1,236 @@
+"""DocumentSource scrapes MDN wiki documents."""
+from __future__ import absolute_import, unicode_literals
+import logging
+
+import dateutil
+
+from .base import DocumentBaseSource
+
+logger = logging.getLogger('kuma.scraper')
+
+
+class DocumentSource(DocumentBaseSource):
+    """Coordinate scraping and local cloning of an MDN Document."""
+
+    OPTIONS = DocumentBaseSource.STANDARD_DOC_OPTIONS
+
+    def load_and_validate_existing(self, storage):
+        """Load the document from storage in simple cases."""
+
+        just_this_doc = (not self.translations and
+                         self.depth == 0 and
+                         self.revisions == 1 and
+                         self.normalized_path)
+        if not self.force and just_this_doc:
+            document = storage.get_document(self.locale, self.slug)
+            if document:
+                return True, []
+        return False, []
+
+    def load_prereqs(self, requester, storage):
+        """Load the data needed for a document."""
+        data = {'needs': []}
+
+        # Load data, gathering further source needs
+        self.load_prereq_normalized_path(storage, data)
+        if self.normalized_path:
+            self.load_prereq_parent_topic(storage, data)
+            self.load_prereq_rendered(storage, data)
+            if data.get('has_rendered'):
+                self.load_prereq_redirect(storage, data)
+            if data.get('is_standard_page'):
+                self.load_prereq_metadata(storage, data)
+                self.load_prereq_english_parent(storage, data)
+                self.load_prereq_history(storage, data)
+                self.load_prereq_children(storage, data)
+
+        return not data['needs'], data
+
+    def load_prereq_normalized_path(self, storage, data):
+        """Load zone data to normalize path, if needed."""
+        if self.normalized_path:
+            return  # Already normalized, done
+
+        # Determine the standard path associated with the zone
+        zone_data = storage.get_zone_root(self.path)
+        if zone_data is None:
+            data['needs'].append(('zone_root', self.path, {}))
+        elif zone_data.get('errors'):
+            raise self.SourceError(
+                'Unable to load zone root for %s', self.path)
+        else:
+            self.normalized_path = self.path.replace(
+                zone_data['zone_path'], zone_data['doc_path'])
+            self.locale, self.slug = self.locale_and_slug(
+                self.normalized_path)
+
+    def load_prereq_parent_topic(self, storage, data):
```

Comment:
👍, and changing to fit the pattern I'm using elsewhere, to avoid too much nesting:

```
if this prerequisite not needed:
   return  # Short explanation

prereq = storage.get_prereq(...)
if prereq is None:
  data['needs'].append(('type_of_prereq', param, {})
else:
  data['key'] = extract_data(prereq)
```